toxqnl/query.c

```diff
@@ -0,0 +1,322 @@
+            break;
+        } else {
+            continue;
+        }
+    }
+
+    freeaddrinfo(root);
+
+    if (info == NULL) {
+        return -1; // No host found
+    }
+
+    const Query new_query = q_make(&ipp, key, name, length);
+
+    // Verify name isn't currently pending response
+    if (q_check(tqnl->pending_list, &new_query, 1) != -1) {
+        return -2; // A similar request is already pending
+    }
+
+    // Send request
+    if (q_send(tqnl->m->dht, &new_query) == 0) {
+        if (q_add(tqnl->pending_list, &new_query)) {
+            return 0;
+        }
+
+        return -3; // Network reports sending error.
+    }
+
+    return -4; // Unknown error.
+}
+
+int query_handle_toxid_response(void *obj, IP_Port source, const uint8_t *pkt, uint16_t length, void *userdata)
+{
+    if (pkt[0] != NET_PACKET_DATA_NAME_RESPONSE) {
+        return -1;
+    }
+
+    if (length <= QUERY_PKT_ENCRYPTED_SIZE + CRYPTO_PUBLIC_KEY_SIZE) {
+        return -1;
+    }
+
+    length -= 1; // drop the packet type
+
+    Tox_QNL *tqnl = (Tox_QNL *)obj;
+
+    // We verify the sender is in our list before we even try to decrypt anything
+    for (size_t i = 0; i < tqnl->pending_list->count; ++i) {
+        if (q_verify_server(&source, &tqnl->pending_list->query_list[i].ipp)) {
+            break;
+        }
+    }
+
+    uint8_t sender_key[CRYPTO_PUBLIC_KEY_SIZE];
+    memcpy(sender_key, pkt + 1, CRYPTO_PUBLIC_KEY_SIZE);
+    length -= CRYPTO_PUBLIC_KEY_SIZE;
+
+    uint8_t nonce[CRYPTO_NONCE_SIZE];
+    memcpy(nonce, pkt + 1 + CRYPTO_PUBLIC_KEY_SIZE, CRYPTO_NONCE_SIZE);
+    length -= CRYPTO_NONCE_SIZE;
+
+    uint8_t clear[length];
+
+    int res = decrypt_data(sender_key, tqnl->m->dht->self_secret_key, nonce,
+                           pkt + 1 + CRYPTO_PUBLIC_KEY_SIZE + CRYPTO_NONCE_SIZE, length, clear);
+
+    if (res == -1) {
+        return -1;
+    }
+
+    Query test = q_make(&source, sender_key, clear, length - CRYPTO_MAC_SIZE);
+
+    int loc = q_check(tqnl->pending_list, &test, 0);
+
+    if (loc < 0) {
+        return -1;
+    }
+
+    if (tqnl->callback) {
+        tqnl->callback(tqnl, tqnl->pending_list->query_list[loc].name, tqnl->pending_list->query_list[loc].length,
+                       clear, userdata);
+    }
+
+    q_drop(tqnl->pending_list, loc);
+    return 0;
+}
+
+Pending_Queries *query_new(void)
+{
+    Pending_Queries *pending_list = (Pending_Queries *)calloc(1, sizeof(Pending_Queries));
+
+    if (!pending_list) {
+        return NULL;
+    }
+
+    pending_list->query_list = (Query *)calloc(1, sizeof(Query));
+
+    if (pending_list->query_list == NULL) {
+        free(pending_list);
+        return NULL;
+    }
+
+    return pending_list;
+}
+
+void query_iterate(Tox_QNL *tqnl)
+{
+    if (tqnl->pending_list->count) {
+        for (size_t i = 0; i < tqnl->pending_list->count; ++i) {
+            q_send(tqnl->m->dht, &tqnl->pending_list->query_list[i]);
```

Comment:
there is no check for the timeout, the query is just sent regardless. this should wait for the next timeout to expire.